bindings/experimental/distrdf/python/DistRDF/Backends/Base.py

```diff
@@ -156,48 +149,61 @@ def mapper(current_range):
             start = int(current_range.start)
             end = int(current_range.end)
 
-            if treename:
+            # Check the type of the headnode. Each headnode will create Range
+            # objects of different kinds and in the distributed mapper will be
+            # needed to create the RDataFrame object according to the constructor
+            # arguments. The following two approaches cannot work
+            # 1. Having separate helper functions to create the RDataFrames,
+            #    for example `rdf = headnode.get_rdf_for_mapper()`, because
+            #    RDataFrame is not serializable (it misses a default constructor)
+            # 2. Checking for the type of the Range object, e.g.
+            #    `if isinstance(current_range, Ranges.TreeRange`, because the
+            #    type of a namedtuple is actually `collections.TreeRange` but
+            #    it's not sufficient to `import collections` at the beginning
+            #    of this file.
+            # For now the easiest and most lightweight approach is to store a
+            # string with a certain name unique to the headnode and use that
+            # as a discriminant.
+            if headnode_type == "TREE":
```

Comment:
This definitely needs further discussion. Ideally we could improve on this by having different `mapper` functions according to the data source so to avoid the `if headnode_type` conditions

Reply:
I see, a main mapper function which is called by the per-source mapper functions? And in the per-source there is only the manipulation of the concrete arguments for that source.

Reply:
Something like that yes, probably better left in another PR. For this PR, either we leave this simple string coming from the HeadNode classes attribute `TYPE` or maybe we can put the same string as an attribute of the `current_range` object

Reply:
why do we need this attribute `TYPE`? doesn't the actual type of the headnode tell you which case we are in?

Reply:
Indeed it does, I could do that for the scope of this refactoring. In the future it might be changed to per-source mappers as hinted above